src/libraries/System.Text.Json/source_generation/src/JsonSourceGenerator.cs

```diff
@@ -0,0 +1,102 @@
+            foreach (CompilationUnitSyntax compilationUnit in receiver.CompilationUnits)
+            {
+                SemanticModel compilationSemanticModel = executionContext.Compilation.GetSemanticModel(compilationUnit.SyntaxTree);
+
+                foreach (AttributeListSyntax attributeListSyntax in compilationUnit.AttributeLists)
+                {
+                    AttributeSyntax attributeSyntax = attributeListSyntax.Attributes.Single();
+                    IMethodSymbol attributeSymbol = compilationSemanticModel.GetSymbolInfo(attributeSyntax).Symbol as IMethodSymbol;
+
+                    if (attributeSymbol?.ToString().StartsWith("System.Text.Json.Serialization.JsonSerializableAttribute") == true)
+                    {
+                        // Get JsonSerializableAttribute arguments.
+                        IEnumerable<SyntaxNode> attributeArguments = attributeSyntax.DescendantNodes().Where(node => node is AttributeArgumentSyntax);
+
+                        int argumentCount = attributeArguments.Count();
+
+                        // Compiler shouldn't allow invalid signature for the JsonSerializable attribute.
+                        Debug.Assert(argumentCount == 1 || argumentCount == 2);
+
+                        // Obtain the one `Type` argument that must be present in the constructor of the attribute.
+                        AttributeArgumentSyntax typeArgumentNode = (AttributeArgumentSyntax)attributeArguments.First();
+                        TypeOfExpressionSyntax typeNode = (TypeOfExpressionSyntax)typeArgumentNode.ChildNodes().Single();
+                        ExpressionSyntax typeNameSyntax = (ExpressionSyntax)typeNode.ChildNodes().Single();
+                        ITypeSymbol typeSymbol = compilationSemanticModel.GetTypeInfo(typeNameSyntax).ConvertedType;
+
+                       // TODO: parse the TypeInfoPropertyName arg if present.
+
+                        Type type = new TypeWrapper(typeSymbol, metadataLoadContext);
+
+                        if (type.Namespace == "<global namespace>")
+                        {
+                            // typeof() reference where the type's name isn't fully qualified.
+                            // The compilation is not valid and user needs to fix their code.
+                            // The compiler will notify the user so we don't have to.
+                            return;
+                        }
+
+                        (SerializableTypes ??= new Dictionary<string, Type>())[type.FullName] = type;
+                    }
+                }
+            }
+
+            if (SerializableTypes == null)
+            {
+                return;
+            }
+
+            Debug.Assert(SerializableTypes.Count >= 1);
+            JsonSourceGeneratorHelper helper = new(executionContext, metadataLoadContext, SerializableTypes);
+            helper.GenerateSerializationMetadata();
+#if LAUNCH_DEBUGGER_ON_EXECUTE
```

Comment:
See other related comment; do we keep this in for Preview releases?

Reply:
Hm my plan was to remove it. I have done this now. Leaving it probably won't be useful to users since they'll be getting a `Release` build.